src/lang/std/artifactGraph.test.ts

```diff
@@ -0,0 +1,743 @@
+      // further more we can check that each edge is bi-directional, if it's not
+      // by checking the arrow heads going both ways, on the graph.
+      await GraphTheGraph(theMap, 2500, 2500, 'sketchOnFaceOnFaceEtc.png')
+    }, 20000)
+  })
+})
+
+function getCommands(codeKey: CodeKey): CacheShape[CodeKey] & { ast: Program } {
+  const ast = parse(codeKey)
+  if (err(ast)) {
+    console.error(ast)
+    throw ast
+  }
+  const file = fs.readFileSync(fullPath, 'utf-8')
+  const parsed: CacheShape = JSON.parse(file)
+  // these either already exist from the last run, or were created in
+  const orderedCommands = parsed[codeKey].orderedCommands
+  const responseMap = parsed[codeKey].responseMap
+  return {
+    orderedCommands,
+    responseMap,
+    ast,
+  }
+}
+
+async function GraphTheGraph(
+  theMap: ArtifactGraph,
+  sizeX: number,
+  sizeY: number,
+  imageName: string
+) {
+  const nodes: Array<{ id: string; label: string }> = []
+  const edges: Array<{ source: string; target: string; label: string }> = []
+  let index = 0
+  for (const [commandId, artifact] of theMap) {
+    nodes.push({
+      id: commandId,
+      label: `${artifact.type}-${index++}`,
+    })
+    Object.entries(artifact).forEach(([propName, value]) => {
+      if (
+        propName === 'type' ||
+        propName === 'codeRef' ||
+        propName === 'subType'
+      )
+        return
+      if (Array.isArray(value))
+        value.forEach((v) => {
+          edges.push({ source: commandId, target: v, label: propName })
+        })
+      if (typeof value === 'string' && value)
+        edges.push({ source: commandId, target: value, label: propName })
+    })
+  }
+
+  // Create a force simulation to calculate node positions
+  const simulation = d3
+    .forceSimulation(nodes as any)
+    .force(
+      'link',
+      d3
+        .forceLink(edges)
+        .id((d: any) => d.id)
+        .distance(100)
+    )
+    .force('charge', d3.forceManyBody().strength(-300))
+    .force('center', d3.forceCenter(300, 200))
+    .stop()
+
+  // Run the simulation
+  for (let i = 0; i < 300; ++i) simulation.tick()
+
+  // Create traces for Plotly
+  const nodeTrace = {
+    x: nodes.map((node: any) => node.x),
+    y: nodes.map((node: any) => node.y),
+    text: nodes.map((node) => node.label), // Use the custom label
+    mode: 'markers+text',
+    type: 'scatter',
+    marker: { size: 20, color: 'gray' }, // Nodes in gray
+    textfont: { size: 14, color: 'black' }, // Labels in black
+    textposition: 'top center', // Position text on top
+  }
+
+  const edgeTrace = {
+    x: [],
+    y: [],
+    mode: 'lines',
+    type: 'scatter',
+    line: { width: 2, color: 'lightgray' }, // Edges in light gray
+  }
+
+  const annotations: any[] = []
+
+  edges.forEach((edge) => {
+    const sourceNode = nodes.find(
+      (node: any) => node.id === (edge as any).source.id
+    )
+    const targetNode = nodes.find(
+      (node: any) => node.id === (edge as any).target.id
+    )
```

Comment:
I don't think you need these `any` type assertions. I know this is just a test, but I'd like us to get out of the habit of disabling the type-checker.

Also, I'm pretty sure this is an actual bug that it's covering up because for all the `edges` populated above, `source` and `target` are strings.

Reply:
Correction: I think the edges type is incorrect.

Reply:
Typing this may be very non trivial but otherwise I agree. @Irev-Dev if you find it tedious maybe we can all contribute to the typing of this.

Reply:
Yeah typing this is going to be a PITA,

It's basically loading extra information onto the nodes and edges so that they can be labelled by plotly, d3 isn't expecting them to be there so it has type errors, but data passes through fine.

I could try and keep track of the labels etc in a separate paired data structure, but I'm not even sure if this is the graphing solution we want, so I'm not sure it's worth the effort, I don' think these current graphs look particularly good, and that might be problem when trying to inspect more complex graphs, an SVG is probably much better since it will make zooming 👌 

Is it okay if this becomes a post-merge thing? because I would rather create better graphs than solve this typeing problem.

Reply:
Sure, if it's still in flux, don't worry about it.